hilti/toolchain/src/compiler/validator.cc

```diff
@@ -204,10 +204,22 @@ struct VisitorPost : visitor::PreOrder, public validator::VisitorMixIn {
                 if ( node->location() == first_node->location() )
                     continue;
 
-                // Functions can legitimately be overloaded.
-                if ( node->isA<declaration::Function>() ) {
-                    if ( first_node->tryAs<declaration::Function>() )
+                // Functions can legitimately be overloaded most of the time.
+                if ( auto current_decl = node->tryAs<declaration::Function>() ) {
+                    if ( auto previous_decl = first_node->tryAs<declaration::Function>() ) {
+                        auto current_fn_ty = current_decl->function()->ftype();
+                        auto previous_fn_ty = previous_decl->function()->ftype();
+
+                        // Overloads cannot differ in return type only.
+                        if ( ! type::same(current_fn_ty->result(), previous_fn_ty->result()) &&
+                             hilti::declaration::areEquivalent(current_fn_ty->parameters(),
+                                                               previous_fn_ty->parameters()) )
+                            error(fmt("'%s' only differs in return type from previous definition in %s", id,
+                                      first_node->location()),
+                                  node);
```

Comment:
WDYT about introducing a dedicated function to check for valid overloads, e.g., a function `bool areOverloads(Function*, Function*)` next to `Function`'s `areEquivalent` which we could use here? Alternative we could introduce a dedicated function `checkOverload` similar to `VisitorMixIn::checkTypeArguments` which encapsulates validation of overloads. Even though the latter might allow better error messages I feel the first one might be better since it could be easier to reuse, e.g., in the resolver.

Reply:
Thanks for the issue, IMO totally fine to do tackle this one problem at a time.